File: Coin Change - Leetcode 322/Coin Change - Leetcode 322.cpp
#include <vector>
#include <algorithm>
using namespace std;


/**
 * Brute force approach with simple recursion
 */
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if (amount == 0)
            return 0;
        else if (amount < 0)
            return -1;

        int min_cnt = -1;
        for (int coin : coins) {
            int cnt = coinChange(coins, amount - coin);
            if (cnt >= 0) 
                min_cnt = min_cnt < 0 ? cnt + 1 : min(min_cnt, cnt + 1);
        }
        return min_cnt;
    }
};



class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount + 1, INT_MAX);
        dp[0] = 0;
        sort(coins.begin(), coins.end());

        for (int i = 1; i <= amount; ++i) {
            for (int coin : coins) {
                if (i - coin < 0) break;
                if (dp[i - coin] != INT_MAX) {
                    dp[i] = min(dp[i], dp[i - coin] + 1);
                }
            }
        }

        return dp[amount] < INT_MAX ? dp[amount] : -1;
    }
};

File: Coin Change - Leetcode 322/Coin Change - Leetcode 322.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # Brute force with simple recursion
        # Time: O(Coins ^ Amount)
        # Space: O(Amount)
        if amount == 0:
            return 0
        elif amount < 0:
            return -1

        min_cnt = -1
        for coin in coins:
            cnt = self.coinChange(coins, amount - coin)
            if cnt >= 0:
                min_cnt = cnt + 1 if min_cnt < 0 else min(min_cnt, cnt + 1)
        return min_cnt



class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # Top Down DP (Memoization)
        # Time: O(Coins * Amount)
        # Space: O(Amount)
        coins.sort()
        memo = {0:0}

        def min_coins(amt):
            if amt in memo:
                return memo[amt]
            
            minn = float('inf')
            for coin in coins:
                diff = amt - coin
                if diff < 0:
                    break
                minn = min(minn, 1 + min_coins(diff))
            
            memo[amt] = minn
            return minn

        result = min_coins(amount)
        if result < float('inf'):
            return result
        else:
            return -1



class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # Bottom Up DP (Tabulation)
        # Time: O(Coins * Amount)
        # Space: O(Amount)
        dp = [0] * (amount + 1)
        coins.sort()

        for i in range(1, amount+1):
            minn = float('inf')
            
            for coin in coins:
                diff = i - coin
                if diff < 0:
                    break
                minn = min(minn, dp[diff] + 1)
            
            dp[i] = minn
        
        if dp[amount] < float('inf'):
            return dp[amount]
        else:
            return -1
                    
